File: cpp/144. Binary Tree Preorder Traversal.cpp
/*
2016-9-30 19:27:31

https://leetcode.com/problems/binary-tree-preorder-traversal/

Given a binary tree, return the preorder traversal of its nodes' values.

For example:
Given binary tree {1,#,2,3},
   1
    \
     2
    /
   3
return [1,2,3].

Note: Recursive solution is trivial, could you do it iteratively?
*/


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> res;
		stack<TreeNode*> s;
		TreeNode* iter = root;
		TreeNode* tmp;
		
		while ((iter != nullptr) || !s.empty())
		{
			if (iter != nullptr)
			{
				res.push_back(iter->val);
				s.push(iter);
				iter = iter->left;
			}
			else
			{
				iter = s.top();
				s.pop();
				
				
				iter = iter->right;
				
			}
		}
		
		return res;
    }
};

File: java/src/main/java/club/cser/leetcode/144. Binary Tree Preorder Traversal.java
package club.cser.leetcode;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

class BinaryTreePreorderTraversal {

    // Morris
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        TreeNode cur = root;

        while (null != cur) {
            if (null != cur.left) {
                TreeNode rightMostNodeOfLeftSubtree = cur.left;

                while (rightMostNodeOfLeftSubtree.right != null && rightMostNodeOfLeftSubtree.right != cur)
                    rightMostNodeOfLeftSubtree = rightMostNodeOfLeftSubtree.right;

                if (rightMostNodeOfLeftSubtree.right == null) {
                    res.add(cur.val);
                    rightMostNodeOfLeftSubtree.right = cur;
                    cur = cur.left;
                } else {
                    rightMostNodeOfLeftSubtree.right = null;
                    cur = cur.right;
                }
            } else {
                res.add(cur.val);
                cur = cur.right;
            }
        }

        return res;
    }

    // iterate
    public List<Integer> preorderTraversalIteration(TreeNode root) {

        List<Integer> res = new ArrayList<>();

        Stack<TreeNode> s = new Stack<>();
        s.push(root);

        while (!s.isEmpty()) {
            TreeNode top = s.pop();

            if (null != top) {
                res.add(top.val);
                s.push(top.right);
                s.push(top.left);
            }
        }
        return res;

    }
}